circuits/circuits/send.circom

```diff
@@ -53,149 +96,80 @@ template Validate() {
+        // Note; users must *not* be allowed to force transactions where the coin values exceed 128 bits and therefore don't pass the range check,
+        // or they'll be able to halt and break the system. TODO: ensure this is enforced by the smart contracts
+        component coin_range_checks[3];
+        for (var i = 0; i < 3; i++) {
+            coin_range_checks[i] = Num2Bits(128);
+        }
+        coin_range_checks[0].in <== highest_coin_to_send;
+        coin_range_checks[1].in <== leaf_coins[0];
+        coin_range_checks[2].in <== leaf_coins[1];
     }
-    coin_range_checks[0].in <== sent_coins[0];
-    coin_range_checks[1].in <== sent_coins[1];
-    coin_range_checks[2].in <== highest_coin;
-    coin_range_checks[3].in <== leaf_coins[0];
-    coin_range_checks[4].in <== leaf_coins[1];
 
     // If the sent coins are valid (i.e., in bounds and in order), then the operator must provide a merkle proof for *some* coins in that range.
     // Since adjacent coin ranges with the same owner are always consolidated, we know that if the sender truly owns all the coins in the sent range,
     // then the leaf coins will owned by them and will be a superset of the spent range.
 
-    // Validate the Merkle proof
-    component leaf = Poseidon(3);
-    leaf.in[0] <== owner;
-    leaf.in[1] <== leaf_coins[0];
-    leaf.in[2] <== leaf_coins[1];
-
-    component merkle_proof_checker CheckMerkleProof(levels);
-    merkle_proof_checker.leaf <== leaf;
-    for (var i = 0; i < levels; i++) {
-        merkle_proof_checker.pathElements[i] <== pathElements[i];
-        merkle_proof_checker.pathIndices[i] <== pathIndices[i];
+    {
+        // Validate the Merkle proof
+        signal initial_root_calculated <== CheckMerkleProof(levels)(
+            leaf <== Poseidon(3)(inputs <== [owner, leaf_coins[0], leaf_coins[1]]),
+            pathElements <== pathElements,
+            pathIndices <== pathIndices
+        );
+        initial_root_calculated === initial_root;
     }
-    merkle_proof_checker.root === initial_root;
-
-    // Require overlap between leaf coins and sent coins, if the sent coins are valid
-    component coin_ranges_overlap = CoinRangesOverlap();
-    for (var i = 0; i < 2; i++) {
-        coin_ranges_overlap.a[i] <== leaf_coins[i];
-        coin_ranges_overlap.b[i] <== sent_coins[i];
-    }
-    (1 - coin_ranges_overlap.out) * sent_coins_valid === 0; // If the sent coins are valid, the sent coins and leaf coins have to overlap
 
     // Check that the leaf coins are owned by the signer and contain the sent coins
-    component leaf_cointains_sent = CoinRangeContains();
-    for (var i = 0; i < 2; i++) {
-        leaf_cointains_sent.superset[i] <== leaf_coins[i];
-        leaf_cointains_sent.subset[i] <== sent_coins[i];
-    }
+    signal is_leaf_cointains_sent <== CoinRangeContains()(
+        set <== leaf_coins,
+        point <== highest_coin_to_send
+    );
```

Comment:
If we make sure that `highest_coin_to_send` is in the Merkle tree, we can actually just force this to be true, which forces the server to provide a Merkle proof for the correct coins:
```
0 === CoinRangeContains()(
    set <== leaf_coins,
    point <== highest_coin_to_send
);
```